CONTRIBUTING.md

```diff
@@ -9,70 +36,187 @@ license the code under the same terms as the project's overall 'outbound'
-may wish to ensure any new code is committed first**.
+Note that the linters *will modify your files* to fix styling errors.
+Make sure that you have saved all your files.
 
-By default, this script checks all files and can take some time; if you alter
-only certain files, you might wish to specify paths as arguments to reduce the
-run-time:
+If you wish to restrict the linters to only the files changed since the last commit
+(much faster!), you can instead run:
 
+```sh
+source ./env/bin/activate
+./scripts-dev/lint.sh -d
 ```
+
+Or if you know exactly which files you wish to lint, you can instead run:
+
+```sh
+source ./env/bin/activate
 ./scripts-dev/lint.sh path/to/file1.py path/to/file2.py path/to/folder
 ```
 
-You can also provide the `-d` option, which will lint the files that have been
-changed since the last git commit. This will often be significantly faster than
-linting the whole codebase.
+## Run the unit tests.
+
+The unit tests run parts of Synapse, including your changes, to see if anything
+was broken. They are slower than the linters but will typically catch more errors.
+
+```sh
+source ./env/bin/activate
+python -m twisted.trial tests
+```
+
+If you wish to only run *some* unit tests, you may specify
+another module instead of `tests` - or a test class or a method:
+
+```sh
+source ./env/bin/activate
+python -m twisted.trial tests.rest.admin.test_room tests.handlers.test_admin.ExfiltrateData.test_invite
+```
+
+If your tests fail, you may wish to look at the logs:
+
+```sh
+less _trial_temp/test.log
+```
+
+## Run the integration tests.
+
+The integration tests run a full Synapse, including your changes, to
+see if anything was broken. They are slower than the unit tests but will
+typically catch more errors.
+
+To run the entire test suite using Sqlite3, use the following command:
+
+```sh
+source ./env/bin/activate
+tox -e py36
+```
+
+If you wish to only run *some* integration tests, you may
+similarly specify a module, e.g.:
+
+```sh
+source ./env/bin/activate
+tox -e py36 -- test.rest.admin.test_room
+```
+
+
+
+Or, to run the test suite using postgres, use the following command. You will need [Docker](https://docs.docker.com/get-docker/) installed.
+
+```sh
+source ./env/bin/activate
+./test_postgresql.sh
+```
+
+
+# 8. Submit your patch.
 
-Before pushing new changes, ensure they don't produce linting errors. Commit any
-files that were corrected.
+Once you're happy with your patch, it's time to prepare a Pull Request.
 
-Please ensure your changes match the cosmetic style of the existing project,
-and **never** mix cosmetic and functional changes in the same commit, as it
-makes it horribly hard to review otherwise.
+To prepare a Pull Request, please:
+
+1. verify that [all the tests pass](#test-test-test), including the coding style;
+2. add a [changelog entry](#changelog);
+3. [sign off](#sign-off) your contribution;
+4. Push code to commits to your fork of Synapse:
+```sh
+git push origin develop
+```
```

Comment:
I would not necessarily recommend pushing to `develop`, even on your own forked repo. It makes it difficult for you to work on a separate set of changes while the first is still being reviewed.

Again, I'm not really sure that we should be trying to teach people how to make PRs here.

Reply:
Fair enough.